Remove react-native-communications

drewvolz opened the pull request:
This PR accomplishes a few things:
* replaces calls for opening mail and phone apps with `openUrl` – docs show this is [built-in to react native](https://reactnative.dev/docs/linking#built-in-url-schemes)
  * fixes #6369 
  * fixes opening `tel:` uri schemes on android
* tests for email and phone number parsing
* removes `react-native-communications` (not been updated since 2017, causing issues)
* removes `@types/react-native-communications`

drewvolz committed:
replace phonecall with openUrl, consolidate prompting logic

drewvolz committed:
replace email with openUrl

drewvolz committed:
uninstall react-native-communications

drewvolz committed:
remove @types/react-native-communications

hawkrives commented on source/components/call-phone.ts:
```diff
@@ -29,3 +32,15 @@ export function callPhone(phoneNumber: string, opts?: Options): void {
 }
+
+const formatNumber = (phoneNumber: string) => {
+	let re = /(\d{3})-?(\d{3})-?(\d{4})/gu
```
Should we handle `+` codes (+1 111 222 3333)?

hawkrives replied:
If not here, we should file an issue to handle it in the future.

drewvolz replied:
We should probably try to handle that in the future. Certain numbers are also showing up as:
- NA
- Without an area code: 7863016

hawkrives commented on source/components/send-email.ts:
```diff
@@ -13,7 +13,7 @@ type Args = {
 	const {to = [], cc = [], bcc = [], subject = '', body = ''} = args
 	try {
-		email(to, cc, bcc, subject, body)
+		openUrl(`mailto:${to}?cc=${cc}&bcc=${bcc}&subject=${subject}&body=${body}`)
```
We need to make sure that we urlencode these segments

drewvolz replied:
Yeah, I agree. I wrote up in the issue that encoding was needed and walked it back incorrectly. I'll try to not muddy the changes too much. Here's at least [how the dependency performed this](https://github.com/anarchicknight/react-native-communications/blob/05bfcdcc123d1e2380156832317a917766699c5c/AKCommunications.js).

hawkrives replied:
I think we can use whatever we use to generate query strings elsewhere, and use it to make this one?

hawkrives replied:
`to=${encodeUriComponent(to)}?{makeMeQueryString({cc, from, subject}, {omitEmpty: true})}` or something like that

drewvolz committed:
encode and querystringify the mailto url components

drewvolz committed:
pass unformatted and url-style numbers to prompt

drewvolz committed:
update formatNumber to handle +1 case

* export formatNumber for testing
* add tests to help us catch regressions
* add regex and matching for different length numbers

drewvolz committed:
* pull out formatEmailParts into its own testable function
* add test cases for email url parts to help catch regressions
* encode to, and query stringify cc, bcc, subject, and body

drewvolz committed:
Merge branch 'master' into drew/remove-rn-communications

hawkrives commented on source/components/__tests__/email.test.ts:
```diff
@@ -0,0 +1,80 @@
+			'mailto:?body=&cc=test%40domain.com&subject=',
+		)
+		expect(formatEmailParts({cc: ['test@domain.com', 'help@domain.com']})).toBe(
+			'mailto:?body=&cc=test%40domain.com&cc=help%40domain.com&subject=',
```
This seems likely problematic. I think we will need to join (but not encode) before we hand it off to openURL?

drewvolz replied:
Could you explain a little more about what the issue is?

drewvolz replied:
I think you're questioning what happens when we do the following 

```ts
let encodedTo = encodeURIComponent(['test@domain.com', 'help@domain.com'].join(','))
let query = querystring.stringify({cc, bcc, subject, body})

return `mailto:${encodedTo}?${query}`
```

specifically the encodedTo bit

```ts
 encodeURIComponent(['test@domain.com', 'help@domain.com'].join(','))
 ```
 
 which looks great according to `openUrl`
 
![telegram-cloud-photo-size-1-5542185941918132147-y](https://user-images.githubusercontent.com/5240843/189504454-7ee343f5-f05e-4442-806c-fc53cd904247.jpg)

hawkrives replied:
I think it works well for "To:" - but we aren't doing the .join for "CC:" and "BCC:". We're adding two `&cc=` parameters, instead.

```
mailto:me@me.me?cc=a@b.com&cc=b@c.com
``` 

I just want to know what happens with that. 

I would expect something more like this:

```
mailto:me@me.me?cc=a@b.com,b@c.com
```

drewvolz replied:
Oh that makes sense. I will fix this.

drewvolz replied:
Okay. So oddly enough the implementation as-is does work for your example cc and bcc addresses.

hawkrives commented on source/components/__tests__/email.test.ts:
```diff
@@ -0,0 +1,80 @@
+			'mailto:?bcc=test%40domain.com&body=&subject=',
+		)
+		expect(
+			formatEmailParts({bcc: ['test@domain.com', 'help@domain.com']}),
```
Same comment as for CC:

hawkrives commented on source/components/call-phone.ts:
```diff
@@ -29,3 +34,28 @@ export function callPhone(phoneNumber: string, opts?: Options): void {
 }
+
+export const formatNumber = (phoneNumber: string): string => {
+	let re = /^(1|)?(\d{3})(\d{3})(\d{4})$/u
```
International numbers can be anything; isn't Great Britain +83?

drewvolz replied:
Yes, I agree there are country calling codes other than +1. 

Phone numbers in other countries don't all have exactly ten digits, though, so the formatting here doesn't apply internationally.

drewvolz replied:
So we're not looking to support +83 by replacing` ^(1|)` with `^(\d|)` in the regex and `'+1 '` with `'+${match[1]} '` in the international code due to the ten digits.

hawkrives replied:
Fair!

drewvolz committed:
remove empty params when creating an email

the spec allows us to append for cc and bcc values as additional params
but it may be simpler for us to handle stripping empty array and string
values from the generated query altogether; at the very least it makes
the tests a little easier to read.
